src/pmi/pmi.cs

```diff
@@ -100,85 +398,323 @@ private static TimeSpan PrepareMethod(Type t, MicroMethodInfo mMI)
+        visitor.FinishType(type);
 
-        ConstructorInfo[] ci = t.GetConstructors(BindingFlagsForCollectingAllMethodsOrCtors);
+        return keepGoing;
+    }
 
-        MicroMethodInfo[] mMI = new MicroMethodInfo[mi.Length + ci.Length];
+    private List<Type> GetInstances(Type type)
+    {
+        List<Type> results = new List<Type>();
 
-        for (int i = 0; i < mi.Length; i++)
+        // Get the args for this generic
+        Type[] genericArguments = type.GetGenericArguments();
+
+        // Only handle the very simplest cases for now
+        if (genericArguments.Length > 2)
         {
-            mMI[i].RMH = mi[i].MethodHandle;
-            mMI[i].name = mi[i].Name;
-            mMI[i].IsAbstract = mi[i].IsAbstract;
-            mMI[i].ContainsGenericParameters = mi[i].ContainsGenericParameters;
+            return results;
         }
 
-        for (int i = 0; i < ci.Length; i++)
+        // Types we will use for instantiation attempts.
+        Type[] typesToTry = new Type[] { typeof(object), typeof(int), typeof(double), typeof(Vector<float>) };
+
+        // To keep things sane, we won't try and instantiate too many copies
+        int instantiationLimit = genericArguments.Length * typesToTry.Length;
+        int instantiationCount = 0;
+
+        foreach (Type firstType in typesToTry)
         {
-            mMI[i + mi.Length].RMH = ci[i].MethodHandle;
-            mMI[i + mi.Length].name = ci[i].Name;
-            mMI[i + mi.Length].IsAbstract = ci[i].IsAbstract;
-            mMI[i + mi.Length].ContainsGenericParameters = ci[i].ContainsGenericParameters;
+            bool areConstraintsSatisfied = AreConstraintsSatisfied(firstType, genericArguments[0]);
+
+            Type secondType = null;
+
+            if (genericArguments.Length == 2)
+            {
+                foreach (Type secondTypeX in typesToTry)
+                {
+                    areConstraintsSatisfied &= AreConstraintsSatisfied(secondTypeX, genericArguments[1]);
+                    secondType = secondTypeX;
+                }
+            }
+
+            if (!areConstraintsSatisfied)
+            {
+                continue;
+            }
+
+            // Now try and instantiate.
+            try
+            {
+                Type newType = null;
+
+                if (genericArguments.Length == 1)
+                {
+                    newType = type.MakeGenericType(firstType);
+                }
+                else if (genericArguments.Length == 2)
+                {
+                    newType = type.MakeGenericType(firstType, secondType);
+                }
+
+                // If we can instantiate, prepare the methods.
+                if (newType != null)
+                {
+                    results.Add(newType);
+                }
+            }
+            catch (Exception)
+            {
+                // Probably missing a constraint check
+            }
```

Comment:
should we log a warning here?

Reply:
Sure.

We should also keep track of types that we fail to instantiate with our simple roster of types, since we won't jit any of those types' methods.

I'll add a note on the latter to the master todo issue.

Reply:
Updated to note types that we can't instantiate and errors when we try and instantiate.